### PR TITLE
Fixed regression: CSS entries were not picked up for storybook pages (e.g. when using exract-text-webpack-plugin)

### LINKED ISSUE #388: extract-text-webpack-plugin with webpack 2
So I am using webpack 2.1.0-beta.21 and storybook 2.3.0 and extract-text-webpack-plugin 2.0.0-beta.3. In full control mode. When I have new ExtractTextPlugin('styles.css') defined things explode.

/node_modules/extract-text-webpack-plugin/index.js:245
                                var shouldExtract = !!(options.allChunks || chunk.isInitial());

if i use extract-text-webpack-plugin 1.0.1 I get this. 
/node_modules/webpack/lib/Chunk.js:37
                throw new Error("Chunk.entry was removed. Use hasRuntime()");

This seems to only be a problem with storybook. It work just fine when running my regular webpack build.

### LINKED ISSUE #708: Error importing scss file
need help with "You may need an appropriate loader to handle this file type."

in compontent file:
`import "../src/components/Header.scss";`

here is my webpack config:

```
const BASE_CSS_LOADER = "css?sourceMap&-minimize";

webpackConfig.module.loaders.push({
  test: /\.scss$/,
  exclude: null,
  loaders: [
    "style",
    BASE_CSS_LOADER,
    "postcss",
    "sass?sourceMap",
  ],
});
webpackConfig.module.loaders.push({
  test: /\.css$/,
  exclude: null,
  loaders: [
    "style",
    BASE_CSS_LOADER,
    "postcss",
  ],
});
webpackConfig.sassLoader = {
  includePaths: project.paths.client("styles"),
};

// File loaders
/* eslint-disable */
webpackConfig.module.loaders.push(
  { test: /\.woff(\?.*)?$/,  loader: "url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff" },
  { test: /\.woff2(\?.*)?$/, loader: "url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff2" },
  { test: /\.otf(\?.*)?$/,   loader: "file?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=font/opentype" },
  { test: /\.ttf(\?.*)?$/,   loader: "url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/octet-stream" },
  { test: /\.eot(\?.*)?$/,   loader: "file?prefix=fonts/&name=[path][name].[ext]" },
  { test: /\.svg(\?.*)?$/,   loader: "url?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=image/svg+xml" },
  { test: /\.(png|jpg)$/,    loader: "url?limit=8192" }
)
/* eslint-enable */

// ------------------------------------
// Finalize Configuration
// ------------------------------------
// when we don't know the public path (we know it only when HMR is enabled [in development]) we
// need to use the extractTextPlugin to fix this issue:
// http://stackoverflow.com/questions/34133808/webpack-ots-parsing-error-loading-fonts/34133809#34133809
if (!__DEV__) {
  debug("Applying ExtractTextPlugin to CSS loaders.");
  webpackConfig.module.loaders.filter((loader) =>
    loader.loaders && loader.loaders.find((name) => /css/.test(name.split("?")[0]))
  ).forEach((loader) => {
    const first = loader.loaders[0];
    const rest = loader.loaders.slice(1);
    loader.loader = ExtractTextPlugin.extract(first, rest.join("!"));
    delete loader.loaders;
  });

  webpackConfig.plugins.push(
    new ExtractTextPlugin("[name].css", {
      allChunks: true,
    })
  );
}

module.exports = webpackConfig;
```

### LINKED ISSUE #1104: storybook didn't load styles
Hi, when i use 
`npm run storybook`
and open browser with http://localhost:6006, I can see the stories that i wrote in my code, but all of the stories has no styles, i checked the page resources with F12, the page didn't load any css file, am i missed some configurations?
```
require('babel-register');
let mainWebpackConfig = require('../webpack/webpack.storybook.config').default;

// Export a function. Accept the base config as the only param.
module.exports = function(storybookBaseConfig, configType) {
  // configType has a value of 'DEVELOPMENT' or 'PRODUCTION'
  // You can change the configuration based on that.
  // 'PRODUCTION' is used when building the static version of storybook.

  let baseLoaders = storybookBaseConfig.module.loaders;
  let customLoaders = mainWebpackConfig.module.loaders;
  storybookBaseConfig.module.loaders = baseLoaders.concat(customLoaders);

  //if not cover the default plugins, it will throw some error message.
  let basePlugins = storybookBaseConfig.plugins;
  let customPlugins = mainWebpackConfig.plugins;
  //remove CaseSensitivePathsPlugin
  basePlugins.splice(3, 1);
  storybookBaseConfig.plugins = basePlugins.concat(customPlugins);
  //for some alias folder
  storybookBaseConfig.resolve = mainWebpackConfig.resolve;

  // storybookBaseConfig.output = mainWebpackConfig.output;
  Object.assign(storybookBaseConfig.entry, mainWebpackConfig.entry);
  // Return the altered config
  return storybookBaseConfig;
};
```
this is my webpack.config.js inside .storybook

```
import ExtractTextPlugin from 'extract-text-webpack-plugin';
import StyleLintPlugin from 'stylelint-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import config from './config';

const paths = config.UTILS_PATH;

const webpackConfig = {
    module: {},
    plugins: [
        new StyleLintPlugin(),
        new ExtractTextPlugin('wealth.css', {
            allChunks: true
        })
    ]
};

const APP_ENTRY_PATH = `${paths.base(config.APP_PATH)}/${config.APP_MAIN}`;
const EXAMPLE_ENTRY_PATH = `${paths.base(config.EXAMPLE_PATH)}/${config.EXAMPLE_MAIN}`;

webpackConfig.entry = {
    wealth: [APP_ENTRY_PATH, EXAMPLE_ENTRY_PATH]
};

webpackConfig.resolve = config.RESOLVE;

webpackConfig.output = {
    filename: `[name].js`,
    path: paths.base(config.BUILD_PATH),
    libraryTarget: 'umd'
};

// Webpack loader settings

let cssLoader = ['css-loader?', 'modules', 'importLoaders=1', 'minimize'];
cssLoader = cssLoader.join('&');

webpackConfig.module.loaders = [{
    test: /\.(js|jsx)$/,
    include: [paths.base(config.APP_PATH), paths.base(config.EXAMPLE_PATH)],
    exclude: /node_modules/,
    loaders: ['babel-loader', 'eslint-loader']
}];

const styleLoaders = [
    {
        test: /\.scss$/,
        include: [paths.base(config.APP_PATH), paths.base(config.EXAMPLE_PATH)],
        exclude: /node_modules/,
        loaders: [
            'style',
            cssLoader,
            'postcss-loader',
            'sass'
        ]
    },
    {
        test: /\.css$/,
        include: [paths.base(config.APP_PATH), paths.base(config.EXAMPLE_PATH)],
        exclude: /node_modules/,
        loaders: [
            'style',
            cssLoader,
            'postcss-loader'
        ]
    }
];

const fontLoaders = [
    {
        test: /\.(woff|woff2|eot|ttf|svg)$/,
        loader: 'url-loader?limit=10000&name=fonts/[name].[ext]'
    }
];

const imageLoader = [
    {
        test: /\.(png|jpg|jpeg|gif|svg)$/,
        exclude: /node_modules/,
        loader: 'url-loader?name=img/img-[hash:6].[ext]&limit=5000'
    }
];

const jsonLoaders = [
    {
        test: /\.json$/,
        loaders: ['json-loader']
    }
];

webpackConfig.module.loaders = [
    ...webpackConfig.module.loaders,
    ...styleLoaders,
    ...fontLoaders,
    ...imageLoader,
    ...jsonLoaders
];

webpackConfig.module.loaders.filter((loader) =>
    loader.loaders && loader.loaders.find((name) => /css/.test(name.split('?')[0]))
).forEach((loader) => {
    const [first, ...rest] = loader.loaders;
    loader.loader = ExtractTextPlugin.extract(first, rest.join('!'));
});

export default webpackConfig;

```
this is my ../webpack/webpack.storybook.config.js

### PR DESCRIPTION
**Pull request by joscha**
Issue: CSS not displaying when using the `exract-text-webpack-plugin`, regression introduced in 38b9dfabfb1b8691c1f7b378f9dfff05b99603f5

## What I did

Upgraded from `@kadira/storybook@1.x` and had the [`extract-text-webpack-plugin`](https://github.com/webpack-contrib/extract-text-webpack-plugin/) enabled.

Unfortunately my CSS didn't load, encountering the same issue as described in #1104 and most likely #388 and #708.

Using the `extract-text-webpack-plugin` creates a new entry point however, making the asset list look something like this:
```js
{ manager: 'static/manager.02f732c9002a59214667.bundle.js',
  preview:
   [ 'static/preview.352040d027a7c22331ae.bundle.js',
     'static/preview.1f62768c19a65e75e4259a4c418d292d.css',
     'static/preview.1f62768c19a65e75e4259a4c418d292d.css.map' ] }
```
and due to https://github.com/storybooks/storybook/blob/34b96aae764a386c35d4db2a057e85236bd8ddfe/app/react/src/server/iframe.html.js#L33 only ever taking the very first asset, the CSS never makes it into `iframe.html`.

## How to test

I added new tests :)

fixes #388 
fixes #708
fixes #1104

### CODE REVIEW

**Comment by joscha** [comment]
@shilman I assigned you, as you seem to be quite active these days :-) Let me know if anything is unclear! Hoping to get the fix for this regression in asap to de-confuse our `extract-text-webpack-plugin`-users :)

**Comment by joscha** [comment]
@UsulPro I saw you were last active, would you mind having a look? I am blocked by this, but need a second pair of eyes on it :)

File: app/react/src/server/iframe.html.js
Comment (shilman) [inline app/react/src/server/iframe.html.js]: who uses this? does it need to be exported?

File: app/react/src/server/iframe.html.js
Comment (shilman) [inline app/react/src/server/iframe.html.js]: u => url?

**Review by shilman — CHANGES_REQUESTED** [review]

File: app/react/src/server/iframe.html.js
Comment (joscha) [inline app/react/src/server/iframe.html.js]: Only the tests, but for those it needs to be exported, yes.

File: app/react/src/server/iframe.html.js
Comment (joscha) [inline app/react/src/server/iframe.html.js]: done in f33b05a6ccb74cae1d681fc3ab7b3ea7

**Review by shilman — APPROVED** [review]

**Comment by joscha** [comment]
Tested in `3.1.7`, works as expected:

```
➜  web git:(joscha/unfork-storybook) yarn build-storybook
yarn build-storybook v0.24.6
@storybook/react v3.1.7

=> Loading custom addons config.
=> Loading custom webpack config (extending mode).
Building storybook ...
ts-loader: Using typescript@2.3.4 and /Users/joscha/Development/canva/web/tsconfig.ts-loader.json
✨  Done in 382.09s.
➜  web git:(joscha/unfork-storybook) ✗ cat target/storybook/iframe.html | grep link
        <link rel='stylesheet' type='text/css' href='preview.89d82dfdc51e6aeaba00b6d8a4785cf3.css'>
```

**Comment by danielmcgrath** [comment]
@joscha do you happen to have a sample project you could point to? I'm on 3.1.7 and can't for the life of me get webpack to actually generate the CSS file. We haven't hit any issues with extract-text-plugin in other projects so I'm wondering if there's a storybook-specific step I'm missing here.

**Comment by joscha** [comment]
@danielmcgrath we use extension mode (e.g. we have an additional `webpack.config.js` in the `conf/` folder and then we do:


```
  const extractCSS = new ExtractTextPlugin({
    // The filename does not contain a hash in dev and local such that conf/head.html used by
    // Storybook can refer directly to the static filename in Storybook.
    filename: options.hashOutputFileNames ? '[name].[contenthash].css' : '[name].css',
    allChunks: true,
    ignoreOrder: true,
  });

// ...
plugins: [
  extractCSS
]
```
that's all that is needed for a `preview.c0ffee.css` to be generated and picked up in the HTML.

**Comment by WellerQu** [comment]
Hi, @joscha , We very love storybook. But we got the error:

```
ERROR in ./src/core/view/style.sass
Module build failed: Error: "extract-text-webpack-plugin" loader is used without the corresponding plugin, refer to https://github.com/webpack/extract-text-webpack-plugin for the usage example
    at Object.module.exports.pitch (/Users/qiuwei/Workspace/rongcaptial-ui/node_modules/extract-text-webpack-plugin/loader.js:25:9)
 @ ./src/core/view/index.js 33:13-36
 @ ./src/core/index.js
 @ ./stories/view.story.js
 @ ./build/config.js
 @ multi ./~/@storybook/react/dist/server/config/polyfills.js ./~/@storybook/react/dist/server/config/globals.js ./~/webpack-hot-middleware/client.js?reload=true ./build/config.js
```

My custorm webconfig is:

```javascript
rules: [
            {
                // use sass-loader for *.sass files
                test: /\.sass/i,
                use: ExtractTextPlugin.extract({ 
                    fallback: 'style-loader',
                    use: [
                        { 
                            loader: `css-loader?modules&localIdentName=${ __DEV__ ? '[path][name]__[local]--[hash:base64:5]' : '[hash:base64]' }`,
                        },
                        {
                            loader: 'postcss-loader',
                            options: {
                                plugins: [
                                    precss,
                                    autoprefixer
                                ]
                            }
                        },
                        {
                            loader: "sass-loader",
                        },
                    ],
                }),
                exclude: /node_modules/
            },
]
```

    "@storybook/react": "^3.1.7", from [here](https://storybook.js.org/basics/slow-start-guide/)
    "webpack": "2.2.0",
    "webpack-merge": "^4.1.0",



Any ideas, please

**Comment by joscha** [comment]
You need to add it to the `plugins` array as well.

**Comment by WellerQu** [comment]
@joscha Yes, you are right. I forgot it. It works now. Thanks.

**Comment by janusch** [comment]
@joscha I somehow cannot get extract styles to work with extending the config based on your example above. Could you expand on your example, or show a whole webpack config.? How do you get the env check into the webpack config when in "extending mode"?
when i run `build-storybook` there are no stylesheets in the output dir, nor is it added to index.html or iframe.html. 
Can you see any oddities or tell me what I am missing? 

 
```
const path = require('path');
const ExtractTextPlugin = require('extract-text-webpack-plugin');

const extractCSS = new ExtractTextPlugin({
  filename: '../build/static/[name].css',
  allChunks: true,
  ignoreOrder: true,
});

module.exports = {
  module: {
    rules: [
      {
        test: /\.md$/,
        use: 'raw-loader',
      },
      {
        test: /\.(jpg|png)$/,
        use: "file-loader"
      },
      {
        test: /\.css$/,
        include: path.resolve(__dirname, '../src/'),
        exclude: path.resolve(__dirname, '../node_modules/'),
        use: extractCSS.extract({
          fallback: 'style-loader',
          use: [
            {
              loader: 'css-loader',
              options: {
                importLoaders: 1,
                modules: true
              },
            },
            {
              loader: 'postcss-loader',
              options: {
                plugins: function(loader) {
                  return [require('autoprefixer')];
                },
              },
            },
          ],
        }),
      },
    ],
  },
  plugins: [
    extractCSS
  ],
};
```

**Comment by ndelangen** [comment]
@janusch Would you be able to try the 3.3 alpha release?

We have changed the way entry points are added there, and the issue you're having should be resolved. could you check it out?

**Comment by janusch** [comment]
@ndelangen Hey, thank you for getting back to me that quickly!
I have just run `build-storybook` with `extract-text-webpack-plugin` and it does now successfully create, copy and include the `preview.css` file.

However, there is another error related to webpack now. Maybe that is due to some other breaking change? It was the first issue in storybook that refers to the same error: https://github.com/storybook-eol/storybook-addons/issues/1 `TypeError: Cannot read property 'emit' of null
    at emitAddReadme` 
I am using the `storybook-readme` addon.

One more little thing is that as soon as I am wrapping the webpack config object in a function returning the object, to get the `env` var, storybook is changing from extended mode to full-control mode. Is there another way to work with an extended config and still make use of `env` for `production` `development`, etc. differentiation?

Thank you for working so hard on storybook, it is really becoming a complex product, with a lot of great features!

**Comment by janusch** [comment]
> One more little thing is that as soon as I am wrapping the webpack config object in a function returning the object, to get the env var, storybook is changing from extended mode to full-control mode. Is there another way to work with an extended config and still make use of env for production development, etc. differentiation?

this one I got resolved!

I am still having the 'emit' error described above. I am running the latest version of the package. `storybook-readme@3.0.6`

Please let me know, if I should test something else, looking forward to get the latest alpha into a stable build asap.

**Comment by janusch** [comment]
I got the remaining issues fixed. Just had to upgrade everything else to 3.3.0-alpha.0 as well.

Thank you for the help, storybook rocks!

**Comment by bsr203** [comment]
sorry I am bit lost as first time using storybook. Do I need a custom webpack conf to load `css`. I have an app based on CRA, and in `stories/index.js`, I am importing the main css file as `import '../index.css';`. I don't see any style applied. 

Uses "@storybook/react": "^3.2.12".

Do I need to use `3.3.0-alpha.0`? I could not see any html file where I can load css. I saw in some older code there was a header html file for customization, but didn't create any with my CRA based app.

thanks for any help.

**Comment by shahankit** [comment]
@shilman  Same issue as @bsr203. I'm using full control mode and I could not find any plugin in default config passed so I added new config using `config.plugins.push(new ExtractTextPlugin('component-styles.[contenthash].css'));` and also update `/\.css$/` to `ExtractTextPlugin.extract`.

**Pull request merged**
